File: api/dl.py
import urllib.error
import urllib.parse
import urllib.request
from http.server import BaseHTTPRequestHandler


class Handler(BaseHTTPRequestHandler):
    def log_message(self, *_args):
        return

    def _cors(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Headers", "X-Canvas-Token, X-Canvas-Base, Content-Type, Accept")
        self.send_header("Access-Control-Allow-Methods", "GET, OPTIONS")

    def do_OPTIONS(self):
        self.send_response(204)
        self._cors()
        self.end_headers()

    def do_GET(self):
        query = urllib.parse.parse_qs(urllib.parse.urlsplit(self.path).query)
        target = query.get("u", [""])[0]
        token = self.headers.get("X-Canvas-Token", "")
        parsed = urllib.parse.urlsplit(target)
        if not target or not token or parsed.scheme not in ("http", "https"):
            self.send_error(400, "Missing or invalid download target")
            return
        try:
            response = urllib.request.urlopen(
                urllib.request.Request(target, headers={"Authorization": "Bearer " + token, "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0 Safari/537.36"}),
                timeout=90,
            )
            body = response.read()
            self.send_response(response.status)
            self.send_header("Content-Type", response.headers.get_content_type() or "application/octet-stream")
            self.send_header("Content-Disposition", "inline")
            self.send_header("Content-Length", str(len(body)))
            self.end_headers()
            self.wfile.write(body)
        except urllib.error.HTTPError as error:
            body = error.read()
            self.send_response(error.code)
            self.send_header("Content-Type", "application/json")
            self.send_header("Content-Length", str(len(body)))
            self.end_headers()
            self.wfile.write(body)
        except Exception as error:
            self.send_error(502, "Download proxy failed: " + str(error))

File: netlify/functions/dl.mjs
const cors = () => ({
  "Access-Control-Allow-Origin": "*",
  "Access-Control-Allow-Headers": "X-Canvas-Token, X-Canvas-Base, Content-Type, Accept",
  "Access-Control-Allow-Methods": "GET, OPTIONS",
  "Access-Control-Max-Age": "86400",
});

export default async (request) => {
  const url = new URL(request.url);
  const target = url.searchParams.get("u") || "";
  const token = request.headers.get("X-Canvas-Token") || "";

  if (request.method === "OPTIONS") return new Response(null, { status: 204, headers: cors() });

  let parsed;
  try { parsed = new URL(target); } catch { parsed = null; }
  if (!target || !token || !parsed || !/^https?:$/.test(parsed.protocol)) {
    return new Response(JSON.stringify({ message: "Missing or invalid download target" }), {
      status: 400,
      headers: { ...cors(), "Content-Type": "application/json" },
    });
  }

  let upstream;
  try {
    upstream = await fetch(target, { headers: { Authorization: "Bearer " + token, "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0 Safari/537.36" } });
  } catch (error) {
    return new Response("Download proxy failed: " + String(error), { status: 502, headers: cors() });
  }

  const out = new Headers(cors());
  out.set("Content-Type", upstream.headers.get("Content-Type") || "application/octet-stream");
  out.set("Content-Disposition", "inline");
  return new Response(upstream.body, { status: upstream.status, headers: out });
};

export const config = { path: "/api/dl" };
